src/mlpack/methods/cf/cf.hpp

```diff
@@ -88,78 +72,90 @@ class CF
    * @param rank Rank parameter for matrix factorization.
+   * @param w First matrix formed after decomposition.
+   * @param h Second matrix formed after decomposition.
+   * @param maxIterations Maximum number of iterations.
+   * @param minResidue Residue required to terminate.
+   * @param mit Whether to terminate only when maxIterations is reached.
    */
-  template<typename FactorizerType = amf::NMFALSFactorizer>
-  CF(const arma::sp_mat& data,
-     FactorizerType factorizer = FactorizerType(),
-     const size_t numUsersForSimilarity = 5,
-     const size_t rank = 0,
-     const typename std::enable_if_t<
-         !FactorizerTraits<FactorizerType>::UsesCoordinateList>* = 0);
+  template<typename mat_type>
+  void Apply(mat_type const& data,
+             const arma::sp_mat& cleanedData,
+             const size_t rank,
+             arma::mat& w,
+             arma::mat& h,
+             const size_t maxIterations,
+             const double minResidue,
+             const bool mit);
 
   /**
    * Train the CF model (i.e. factorize the input matrix) using the parameters
    * that have already been set for the model (specifically, the rank
-   * parameter), and optionally, using the given FactorizerType.
+   * parameter), and optionally, using the given DecompositionPolicy.
    *
-   * @param data Input dataset; coordinate list or dense matrix.
-   * @param factorizer Instantiated factorizer.
+   * @param data Input dataset; dense matrix (coordinate lists).
+   * @param decomposition Instantiated DecompositionPolicy object.
+   * @param maxIterations Maximum number of iterations.
+   * @param minResidue Residue required to terminate.
+   * @param mit Whether to terminate only when maxIterations is reached.
    */
-  template<typename FactorizerType = amf::NMFALSFactorizer>
   void Train(const arma::mat& data,
-             FactorizerType factorizer = FactorizerType());
+             DecompositionPolicy decomposition = DecompositionPolicy(),
+             const size_t maxIterations = 1000,
+             const double minResidue = 1e-5,
+             const bool mit = false);
 
   /**
    * Train the CF model (i.e. factorize the input matrix) using the parameters
    * that have already been set for the model (specifically, the rank
-   * parameter), and optionally, using the given FactorizerType.
+   * parameter), and optionally, using the given DecompositionPolicy.
    *
-   * @param data Sparse matrix data.
-   * @param factorizer Instantiated factorizer.
+   * @param data Input dataset; sparse matrix (user item table).
+   * @param decomposition Instantiated DecompositionPolicy object.
+   * @param maxIterations Maximum number of iterations.
+   * @param minResidue Residue required to terminate.
+   * @param mit Whether to terminate only when maxIterations is reached.
    */
-  template<typename FactorizerType = amf::NMFALSFactorizer>
   void Train(const arma::sp_mat& data,
```

Comment:
Perhaps I missed something, but we should at least provide a Train method, that does use the decomposition method from the constructor, let me know what you think.

Reply:
You are right, we already know the decomposition policy at construction time.